Scripts/Runtime/Entities/Lifecycle/Spawning/EntitySpawnSystem.cs

```diff
@@ -0,0 +1,424 @@
+        //*************************************************************************************************************
+
+        /// <summary>
+        /// Spawns an <see cref="Entity"/> with the given definition immediately by cloning the passed in prototype
+        /// <see cref="Entity"/> and returns it immediately. 
+        /// </summary>
+        /// <remarks>
+        /// This will not enable this system.
+        /// </remarks>
+        /// <param name="prototype">The prototype <see cref="Entity"/> to clone</param>
+        /// <param name="spawnDefinition">
+        /// The <see cref="IEntitySpawnDefinition"/> to populate the created <see cref="Entity"/> with.
+        /// </param>
+        /// <param name="shouldDestroyPrototype">
+        /// If true, will destroy the prototype <see cref="Entity"/> after creation.
+        /// </param>
+        /// <typeparam name="TEntitySpawnDefinition">The type of <see cref="IEntitySpawnDefinition"/></typeparam>
+        /// <returns>The created <see cref="Entity"/></returns>
+        public Entity SpawnImmediate<TEntitySpawnDefinition>(Entity prototype, TEntitySpawnDefinition spawnDefinition, bool shouldDestroyPrototype = false)
+            where TEntitySpawnDefinition : unmanaged, IEntitySpawnDefinition
+        {
+            EntityPrototypeSpawner<TEntitySpawnDefinition> entitySpawner = GetOrCreateEntitySpawner<EntityPrototypeSpawner<TEntitySpawnDefinition>, TEntitySpawnDefinition>();
+            return entitySpawner.SpawnImmediate(prototype, spawnDefinition, shouldDestroyPrototype);
+        }
+
+        //TODO: Implement a SpawnImmediate that takes in a NativeArray or ICollection if needed.
+
+        private void EnableSystem(IEntitySpawner entitySpawner)
+        {
+            //By using this, we're writing immediately, but will need to execute later on when the system runs
+            Enabled = true;
+            m_ActiveEntitySpawners.Add(entitySpawner);
+        }
+
+        protected override void OnUpdate()
+        {
+            Dependency = ScheduleActiveEntitySpawners(Dependency);
+            
+            //Ensure we're turned back off
+            m_ActiveEntitySpawners.Clear();
+            Enabled = false;
+        }
+
+        private JobHandle ScheduleActiveEntitySpawners(JobHandle dependsOn)
+        {
+            NativeArray<JobHandle> dependencies = new NativeArray<JobHandle>(m_ActiveEntitySpawners.Count + 1, Allocator.Temp, NativeArrayOptions.UninitializedMemory);
+            dependencies[^1] = m_EntityArchetypes.AcquireAsync(AccessType.SharedRead, out NativeParallelHashMap<long, EntityArchetype> entityArchetypesLookup);
+
+            //All the active spawners that need to go this frame are given a chance to go ahead and run their spawn jobs
+            int index = 0;
+            foreach (IEntitySpawner entitySpawner in m_ActiveEntitySpawners)
+            {
+                EntityCommandBuffer ecb = m_CommandBufferSystem.CreateCommandBuffer();
```

Comment:
We could, but I opted to do it this way because of the need to to "complete" the ECB at the end of the block. 

You create an ECB and then you have to add the job handle for producer to the system. 

With it created here, you see the creation and the addition at the end for the amalgamated jobhandle in the same spot. 

If we created inside, we'd have to add for producer inside for each spawner (inefficient) or keep it outside but then maybe worry we weren't adding it when just looking at the inner code.

Reply:
oh ok, the AddJobHandleForProducer call at the end. Just a comment to clarify then. Thanks!

Reply:
Done